### DISCUSSION TITLE
Creating a simple INSERT policy not allowing me to insert

### DISCUSSION THREAD

**Discussion by guicaro** [supabase/supabase#463] [Questions] [answered]
## Description

I am testing out policies and have created a simple table with following schema:

```
id - integer
name - text
```

The policy I created only for INSERT is:

```
WITH CHECK:
(name = ANY (ARRAY['yoda'::text, 'obiwan'::text]))
```

Note that my original expression was `(name = IN ('yoda', 'obiwan'))` but supabase did some kind of rewriting. Now, when trying to issue the following request in bash I get an error:

```
curl -X POST URL_TO_MY_DB_AND_TABLE \
-H "apikey: SUPABASE_KEY" \
-H "Authorization: Bearer SUPABASE_KEY" \
-H "Content-Type: application/json" \
-d '{ "name": "yoda"}'
```

## Error
```
{"hint":null,"details":null,"code":"42501","message":"new row violates row-level security policy for table \"testingpolicies\""}
```

If I disable policies on that table, the INSERT statement in bash works fine. Am I missing something?

**Comment by guicaro** [comment]
even if I add a simpler expression (which always evaluates to TRUE), I still can't seem to insert a row 🤔. By the way, I am using the _anon_, _public_ API Key. Maybe I need to use a different key? :

```
WITH CHECK:
(1 = 1)
```

**Comment by kiwicopple** [comment]
Hey @guicaro - sorry to hear you're having trouble with RLS. Could you please give some steps to replicate this?

Judging by the info you sent, have you done both of these on your table?

```sql
alter table public.testingpolicies enable row level security;
create policy "Allow insert access" on public.testingpolicies for insert with check ( true );
```

**Reply by guicaro** [reply]
Hi @kiwicopple ,  thanks for the prompt reply. I was doing a quick test using the customized bash scripts that supabase provides in the API section of dashboard (really nice attention to detail and great way to get up and running). Today I was playing with the supabase client inside my ReactJS application. I am able to read and insert to a table with no problem. But I did notice that when I activate RLS on the table I keep getting the same error. I even tried adding the `{returning: 'minimal'}' argument as described in [this comment](https://github.com/supabase/supabase/discussions/270#discussioncomment-235033), but no luck

<img width="590" alt="Screen Shot 2021-01-02 at 11 56 14 PM" src="https://user-images.githubusercontent.com/1750596/103472769-0f5a4600-4d57-11eb-86ef-0e14406626f6.png">

Also, does clicking on the lock button in `Authentication -> Policies` not enable RLS on the table? Does one still have to issue the `alter table` statement you mentioned? If that is the case, what is this lock button for?
<img width="157" alt="Screen Shot 2021-01-03 at 12 04 52 AM" src="https://user-images.githubusercontent.com/1750596/103472808-7415a080-4d57-11eb-90c9-6813f475fb6e.png">

I've been using the UI to create the policies, are there any current issues with it? Should one issue the `create policy` statement instead?

Some context. I am looking into implement a form of rate limit via RLS, whereby an insert is not allowed if a condition is not met.

**Reply by kiwicopple** [reply]
Hey @guicaro sorry I missed this one.

> I keep getting the same error.

Can you give me some code to replicate this? Maybe a table + policy script?

> Also, does clicking on the lock button in Authentication -> Policies not enable RLS on the table?

Yes it does - you don't need to use the script and the interface is fine. It's just clearer for me to add code snippets. If the padlock is "red", RLS is disabled. if it's "white" then RLS is enabled

**Comment by eden-lane** [comment]
I'm also trying to add simple INSERT policy:
```
user_id = auth.uid()
```

It fails because user_id cannot be null. Am I missing something?

**Reply by eden-lane** [reply]
Okay, I managed to do it, but record has no user_id for some reason

**Reply by guicaro** [reply]
@eden-lane can you expand on steps you went through to create your INSERT policy and still be able to insert a record that evaluates this policy? Anything in particular that helped you?

**Reply by joshnuss** [reply]
@eden-lane make sure your table has default value for the `user_id` column. ie:

```sql
create table rocket_ships (
  id serial primary key,
  user_id uuid references auth.users not null default auth.uid()
)
```

**Comment by carlesjm** [comment]
Hi all. I'm trying to do the same and  testing RLS. I've created a simple table with row level security enabled with just one policy for insert to enable anyone to insert (with check true). But I get the same error. Imagine the case that anyone can insert to a table but only some users can read this information. An anonymous user should be able to insert a new row but I can't manage to do this.

Should I add some other policy for select?

**Reply by kiwicopple** [reply]
Hey @carlesjm - let us know if you have some code to replicate this? This should work for you

```sql
create table test (
  id bigint GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  name text
);
alter table test enable row level security;
create policy "Allow insert access" on public.test for insert with check ( true );
```

After doing that you can do 

```js
const { data, error } = supabase.from('test').insert({ name: "hello" })
```

**Reply by carlesjm** [reply]
Hi @kiwicopple and thanks for your reply. I had the same code as you suggest (with more columns, but essentialy the same). I've tried it with exactly your code now with supabase client and bash, but I get the same result:

`{"hint":null,"details":null,"code":"42501","message":"new row violates row-level security policy for table \"test\""}`

Does this code work for you?

**Comment by steve-chavez** [comment, marked as answer]
@guicaro Can you try adding the `Prefer: return=minimal` header to your request. Like:

```bash
curl -X POST URL_TO_MY_DB_AND_TABLE \
-H "apikey: SUPABASE_KEY" \
-H "Authorization: Bearer SUPABASE_KEY" \
-H "Content-Type: application/json" \
-H "Prefer: return=minimal" \
-d '{ "name": "yoda"}'
```

The problem is that by default, PostgREST gets the primary key after a successful INSERT([docs](https://postgrest.org/en/v7.0.0/api.html#insertions-updates)) and this also needs a SELECT privilege/policy. `return=minimal` supresses that.

**Reply by carlesjm** [reply]
This worked for me. Thanks!

Does the client implement this header?

**Reply by steve-chavez** [reply]
@carlesjm Yes, this feature was added not too long ago on https://github.com/supabase/postgrest-js/pull/133.

You can use it like:

```js
await supabase
    .from('users')
    .insert({ username: 'bar' }, { returning: 'minimal' })
```

**Reply by carlesjm** [reply]
Great! Thank you very much @steve-chavez !

**Reply by LeoHeuser** [reply]
Hey, I think this is also the solution for my problem. I just want to write in a database; and no one can read from it. I developing in Swift. How do I pass the `returning:“minimal“` in Swift? I tried some stuff but nothing worked so far.

Thx ✌🏼

**Reply by steve-chavez** [reply]
@LeoHeuser Maybe try asking for the feature request on https://github.com/supabase-community/postgrest-swift.
